### PR TITLE
test: fix udp-multicast-join tests

### PR DESCRIPTION
**Pull request by santigimeno**
The messages must be actually sent to the multicast address.

### CODE REVIEW

**Comment by santigimeno** [comment]
Unfortunately I've not been able to make the ipv6 tests to pass on `smartos`, `aix`, `zos` and `s390x`. I'd need some help with those.

CI: https://ci.nodejs.org/job/libuv-test-commit-linux/1270/

**Comment by bnoordhuis** [comment]
cc @libuv/smartos and @mhdawson for the other three platforms.

**Comment by santigimeno** [comment]
Just an update: `zos` is Ok.
The `ipv6` test is failing on `smartos` and `aix`.
The `ipv4` test is failing on linux `rhel72-s390x`.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1242/

cc @libuv/aix

**Comment by mhdawson** [comment]
@gireeshpunathil would you be able to take a look at the aix issue?

**Comment by mhdawson** [comment]
@miladfarca could you look at the rhel 390 issue?  If you need access to the specific machine just let me know.

**Comment by santigimeno** [comment]
Thanks @mhdawson

**Comment by miladfarca** [comment]
@mhdawson @santigimeno Problem had to do with iptables (firewall). Multicast packets were being received by `eth0` but OS was rejecting them. Adding the following to `/etc/sysconfig/iptables` and restarting iptables fixed the issue and the test runs fine. This rule can be further customized to narrow the accepted list.
```
-A INPUT -m pkttype --pkt-type multicast -j ACCEPT
```

**Comment by santigimeno** [comment]
@miladfarca thanks for looking into it!
@mhdawson, @refack would it be possible adding that rule (or similar) to the `rhel72-s390x` box in the CI?

**Comment by cjihrig** [comment]
Ping @mhdawson and @refack regarding the last comment ^

**Comment by refack** [comment]
We could do that... The question is; is it better to patch the system or the test? AFAIK we do minimal changes to iptables so our setup should be a good canary for typical LinuxONE setups...

I don't have a good answer, on the one hand we want to cover the code, on the other, we want to be aware of possible regressions/issues with typical systems.

I defer to the IBM team.

**Comment by cjihrig** [comment]
> The question is; is it better to patch the system or the test?

IMO, it would be better to change the test if that's possible. Otherwise, we'll probably end up with users opening issues about the test.

**Comment by mhdawson** [comment]
@refack I think the ansible config for linuxOne may already be making some IPTable additions in which case this would be a natural extension. @sam-github can you look to see if I'm remembering correctly and if so create a PR to add the additional configuration?

**Comment by sam-github** [comment]
See https://github.com/nodejs/build/pull/1808

**Comment by santigimeno** [comment]
I just run the CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1452/ and it does not look good.

- `rhel72-s390x` [still failing](https://ci.nodejs.org/job/libuv-test-commit-linux/1497/nodes=rhel72-s390x/) the `ipv4` after apparently allowing multicast in the CI bots.
- `smartos` and `aix` still failing the `ipv6` test. I spent some time trying to make them pass in `smartos` but could not figure it out.
- `fedora-last-latest-x64` both tests fails. This one is new, so I don't really know :man_shrugging:. I'll try look into it.

/cc @libuv/aix @libuv/smartos @mhdawson @miladfarca 

> IMO, it would be better to change the test if that's possible

The tests themselves are quite simple. TBH I don't know what could I change. Maybe disabling them for the platforms we can't make them pass is an option :man_shrugging: .

**Comment by bnoordhuis** [comment]
I'll take a look at AIX later this week. @cjihrig Perhaps you do smartos?

**Comment by miladfarca** [comment]
I just ran the test manually on `rhel72-s390x-1` and it passed. iptables is configured correctly. I don't seem to have access to `rhel72-s390x-2`, can my key get added to the machine or could you check if https://github.com/libuv/libuv/pull/2185#issuecomment-467290846 is applied on this machine, my pubic key if needed:
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDQZihjEXcY52UZo09CEb29HwOWwXcmwbwZFR4rsACQQyGUexL4fkVrFfwuG8eV1vg5KECsO8LiGY/MHkNIpABaJJoip0Qxgv0pAFtAukIDjLXXOV/VNJjfIto16vOAehRZkmI+BtQP8TjoT2CSyJgvVQcay8BhH52in1LQQsyCi2crHLYzDrrCgY/rAmuVb1MzMnT8mFOdJ8E5RBhjnmc1K4YBKmNTf6yefgbOJssI0lLzp7Q2uytzp3pipg7AO/VqmRn8953UTJS/cOQeBi3nCYGpz4I7kOKHgwbdW1IP/XFfm0KO5daulHQeToRGIE85ntxF314wsYE3ZyeJwKgH
```

**Comment by mhdawson** [comment]
@sam-github can you add @miladfarca's key?

**Comment by sam-github** [comment]
Key added to test-linuxonecc-rhel72-s390x-1

**Comment by miladfarca** [comment]
@sam-github Need access to `rhel72-s390x-2` if its ok, already had access to `-1`. thanks again.

**Comment by sam-github** [comment]
@miladfarca test-linuxonecc-rhel72-s390x-2 has your keys now

**Comment by miladfarca** [comment]
thanks @sam-github , issue is solved and tests are passing on both of our `rhel72-s390x` machines now.

**Comment by sam-github** [comment]
Great, thanks. I can't close this, maybe someone else here can.

**Comment by mhdawson** [comment]
@miladfarca what needed to be updated?  Just want to confirm that it is covered by the current Ansible scripts.

**Comment by miladfarca** [comment]
@mhdawson `/etc/sysconfig/iptables` had to be updated and this rule was added:
```
-A INPUT -m pkttype --pkt-type multicast -j ACCEPT
``` 
Seems like sam had opened a card for this: https://github.com/nodejs/build/pull/1808

**Comment by cjihrig** [comment]
@santigimeno I think the IPv6 test might need some additional skipping logic. At least on SmartOS, I think the test requires at least one interface that is UP, MULTICAST, IPv6, not LOOPBACK, and not POINTOPOINT. By passing the `can_ipv6()` check in the test, we have established UP and IPv6. We could add a check for `is_internal == 0` as well to cover the `not LOOPBACK` case and that would probably be adequate. I don't think the libuv public API exposes the MULTICAST or POINTOPOINT flags. We could work around that in libuv by calling `getifaddrs()` directly, but it could present an issue in projects like Node if the same logic is necessary.

TL;DR - I think if we can get the IPv6 test passing by adding just adding an `is_internal` check, then let's do that. If it still fails, I'd be OK with just skipping it completely on SmartOS.

**Comment by santigimeno** [comment]
@cjihrig Thanks for the info! I've been able to confirm your assesment by adding a MAC-derived ipv6 to my instance as described [https://wiki.smartos.org/pages/viewpage.action?pageId=4554859](here) and now the test pass. I'll see about adding the `is_internal` check, but I think I can also be nice trying to add a not`LOOPBACK` ipv6 to the `smartos` CI machines.

**Comment by santigimeno** [comment]
I've been looking into the failures in the `fedora-last-latest-x64` bot and the problem seems to be that `firewalld` is installed and running and has a configuration that doesn't allow any kind of external incoming traffic apart from `ssh`, `mdns` and `dhcp`. The `multicast` tests fail because though run locally, they use the 'external' network interface and not the `loopback`.

To make it work I've had to make the following changes to the firewalld configuration:
```sh
firewall-cmd --permanent --direct --add-rule ipv4 filter INPUT 0 -m pkttype --pkt-type multicast -j ACCEPT
firewall-cmd --permanent --direct --add-rule ipv6 filter INPUT 0 -m pkttype --pkt-type multicast -j ACCEPT
```
Also we should disable the `IPv6_rpfilter` configuration in `/etc/firewalld/firewalld.conf` as it's dropping the ipv6 datagrams.

Another option might be not using `firewalld` altogether, which other bots seems not to be using at all.

Any thoughts on whether these changes could be applied? /cc @mhdawson @sam-github @refack

**Comment by mhdawson** [comment]
@santigimeno the firewall-cmd commands or the commands to disable firewalld altogether should be added to the ansible templates so that they run when fedora machines are configured. Look for  "- name: Firewall"https://github.com/nodejs/build/blob/master/ansible/roles/jenkins-worker/tasks/main.yml as an example of what was done on rhel72-s390x.  It might even be possible that simply expanding the condition so that those run on fedora as well solves the problem but you'd have to review them to see.

**Comment by santigimeno** [comment]
Added `can_ipv6_external()` helper method as suggested in https://github.com/libuv/libuv/pull/2185#issuecomment-509769762.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1529/

**Comment by santigimeno** [comment]
The CI looks pretty good now. There's only the issue with `centos7-64` that should be solved once https://github.com/nodejs/build/pull/1879 lands. So I'd say the PR is finally ready.

**Comment by santigimeno** [comment]
@bnoordhuis after [this patch](https://github.com/libuv/libuv/pull/2185/commits/38f7c7b59ad3cdc5b6a44a9394fd2f0620bcfa94), that tried to fix the `smartos` failures as suggested by [Colin](https://github.com/libuv/libuv/pull/2185#issuecomment-509769762), the aix bots are also happy. Is this ok from the aix pov?

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/b571851e6e3040d009a971f0bebe010426bfcf87. Thanks!
